src/freenas/etc/rc.conf

```diff
@@ -84,3 +84,6 @@ inadyn_flags="--continue-on-error"
 
 # Apply CPU microcode updates, if any
 microcode_update_enable="YES"
+
+# Let's increase timeout for rc.shutdown(8) to 5 minutes
+rcshutdown_timeout=310
```

Comment:
It is `90` by default and you are adding up to 300 seconds of tasks so it should probably become 390.

Reply:
I don't think it adds it ?
```
(int) If set, start a watchdog	timer in the background	which
		 will terminate	rc.shutdown if shutdown(8) has not completed
		 within	the specified time (in seconds).  Notice that in addi-
		 tion to this soft timeout, init(8) also applies a hard	time-
		 out for the execution of rc.shutdown.	This is	configured via
		 sysctl(8) variable kern.init_shutdown_timeout and defaults to
		 120 seconds.  Setting the value of rcshutdown_timeout to more
		 than 120 seconds will have no effect until the	sysctl(8)
		 variable kern.init_shutdown_timeout is	also increased.
```

Reply:
Assuming normal system shutdown takes 90 seconds (maximum allowed by default timeout) and you are also adding additional shutdown tasks that last 300 seconds (again, maximum allowed by our chosen timeout), you'll need this timer to be 90 + 300 for both operations to complete successfully.

Reply:
If shutdown scripts takes 300 seconds as it is allowed, you'll only have 10 seconds left for all other services which is not good.

Reply:
> Assuming normal system shutdown takes 90 seconds
@themylogin It has been like this before. So before it was only 90 seconds for both normal shutdown operations and the init tasks we had defined. So umm I can surely increase it to 390 but would we like that ?

Reply:
What is the point of increasing it to 310 then?

Reply:
Reasoning behind that is that at 300 seconds init and the watchdog both will force down a hard limit on the timeout resulting in the system thinking the rc.shutdown script errored out. So yes it did not timeout as desired, but at least if we are the ones running the tasks, we can make that clear and timeout our tasks at exactly 300. However if there are any system related tasks still running, they will get errored out at 310 and we would know that it was not us which resulted in the issue.

Reply:
I mean what's the point of increasing it from 310 if it was 90 before?

Reply:
Because if we don't do that, watchdog will timeout rc.shutdown at 90 secs and as we would like to increase the timeout for shutdown, that is not desirable.

Reply:
Imagine if some script that is executed before ix-shutdown takes 20 seconds which is perfectly normal, will your logic in https://github.com/freenas/freenas/pull/2596#discussion_r258388387 still work?

Reply:
I think I should rephrase this. 300 is the maximum time we allow for all shutdown related tasks controlled by rc.shutdown. Yes you are right that the statement would not hold to this fact if there are other system utilities being executed in the same duration as well. But I don't think there are any other except from out ix-shutdown script ?
```
   Operation of rc.shutdown
     1. Source /etc/rc.subr to load various rc.subr(8) shell functions to use.
     2. Load the configuration files.
     3. Invoke rcorder(8) to order the files in /etc/rc.d/ and the $local_startup directories that have a ``shutdown'' KEYWORD (refer to rcorder(8)'s -k flag), reverse that order,	and assign the result to a variable.
     4. Call each script in turn using run_rc_script() (from rc.subr(8)), which sets $1	to ``stop'', and sources the script in a subshell.  If the script has a .sh suffix then it is sourced directly into the current shell.
``` 

I have no issues increasing the timeout, please let me know if you think we still should ?

Reply:
> 300 is the maximum time we allow for all shutdown related tasks controlled by rc.shutdown

Than you should time out ix-shutdown at 210 seconds because all other system tasks should still have 90 seconds to execute

Reply:
There are no other system tasks being controlled by rc.shutdown except for ix-shutdown right now if i am not mistaken ? So that means it only has ix-shutdown to execute ?

Reply:
There are dozens of scripts that run at shutdown.

There isnt much sense to turn this into a job, we can simply use midclt -t XXX call foo.bar.

I wasnt aware default shutdown was set to 90 by default and I am not sure anymore increasing it for ix-shutdown would make much sense.

Perhaps rcshutdown_timeout should be a dynamic value based on some conditions (e.g. number of shutdown scripts [perhaps even configurable timeout per script], number of VMs [we have to wait for them to cleanly shutdown])

Its a bit more complicated than I previously anticipated ; )

Reply:
Oh we also have to make sure ix-shutdown depends on middlewared now.

And if choose to go the timeout time per script way it would have to be a job.

Reply:
I like both William's ideas

Reply:
@william-gr @themylogin agreed :)